### ISSUE TITLE
Exclude a specific Python class

### ISSUE THREAD

**Issue by hadim** [microsoft/pyright#943]
Would that be possible to exclude a specific Python class/object.

I extensively use [omegaconf](https://omegaconf.readthedocs.io/en/latest/index.html) that provides a dynamic dict `omegaconf.Omegaconf` and pyright can't detect the keys in it are set and so things like the following:

```python
config = omegaconf.Omegaconf.load("config.yaml")
config.report_name = "hello"
```

shows an error `Cannot access member "report_name" for type "OmegaConf"`.

**Comment by erictraut** [comment]
Are you using the latest version of omegaconf? I just installed it, and I'm not seeing the error you're seeing. That's because the `load` method returns a value of type `Union[DictConfig, ListConfig]`, and both of those classes define a `__getattr__` method that returns type `Any` and a `__setattr__` that accepts an input parameter of type `Any. Pyright will therefore allow you to access or set attributes of any name without emitting an error.

**Comment by hadim** [comment]
Thank you for looking into this. It looks like indeed pyright is doing the right thing and my minimal example wasn't reproducing my errors:

```python
spec: omegaconf.OmegaConf = omegaconf.OmegaConf.load("config.yaml")
spec.metadata.eeee = 'dsdsd'
```

A put the type of my objects to `omegaconf.OmegaConf` everywhere, I think this is why I have the error.

Amnther one too:

```python
data_dict: Union[dict, omegaconf.OmegaConf]
data_dict = {"key1": 55, "key2": "dsdsd"}
spec: omegaconf.OmegaConf = omegaconf.OmegaConf.create(data_dict)
```

**Comment by hadim** [comment]
Another one:

```python
spec: omegaconf.OmegaConf = omegaconf.OmegaConf.load("config.yaml")
spec.metadata.eeee = 'dsdsd'
omegaconf.OmegaConf.save(spec, "my/path")
```

Another one:

```python
spec: omegaconf.OmegaConf = omegaconf.OmegaConf.load("config.yaml")
spec.metadata.eeee = 'dsdsd'
print(spec.pretty())
```

I am using the latest version of Pylance.

**Comment by erictraut** [comment]
Yes, the type annotation that you've added (`omegaconf.OmegaConf`) is incorrect in this context, and Pyright is correctly pointing out that the type of the RHS expression is not assignable to this type.

If you simply remove this annotation, Pyright will properly infer the type of the variable `spec`.

**Comment by hadim** [comment]
So I should always use `DictConfig` instead of `omegaconf.OmegaConf`?

**Comment by erictraut** [comment]
I don't know anything about this library, but based on the type annotations within the library code, it looks like the `create` method returns either a `DictConfig` or a `ListConfig` or a union of these two types. Several overloads are provided, so the type checker will be able to figure out whether it returns a `DictConfig` or a `ListConfig` in most cases based on the input parameters.

Is there a reason why you are annotating `spec`? I typically don't annotate my local variables unless the type checker is unable to infer the type from the value that is assigned to it.

**Comment by hadim** [comment]
In fact I showed you simple code but all of this is within functions/methods. Here is another one (but I guess this one is a more general question):

```python
def set_template(self, template_spec: Union[dict, omegaconf.DictConfig, omegaconf.ListConfig]):
    template_spec = omegaconf.OmegaConf.create(template_spec)
```

or the equivalent using `pathlib`:

```python
def get_file(filepath: os.PathLike):
    filepath: pathlib.Path = pathlib.Path(filepath)
    print(filepath.is_file())
```

The idea is to allow multiple "similar" types as input in a single argument but fix it at the beginning of the method to a type I control and I prefer to work with such as `pathlib.Path` and `DictConfig`.

**Comment by erictraut** [comment]
Type annotations don't perform any type of "casting" nor do they have any effect at runtime. They simply declare that a symbol will contain values of a particular type.

Functions that accept union types generally need to include runtime tests to disambiguate between different types. You will often see calls to the `isinstance` method.

```
def foo(param: Union[int, str]):
    if isinstance(param, str):
        # Pyright knows param must be str at this point
        do_something_with_str(param)
    else:
        # Pyright knows param must be an int at this point
        do_something_with_int(param)
```

**Comment by hadim** [comment]
Thanks for your time @erictraut
